### PR TITLE
Add databuilder

### LINKED ISSUE #803: Add data builder
Add a data builder to support data transformation from one stage to another. 

DataBuilder is used to manipulate and store data. The data from the previous filter can be transformed and stored in the context so that the data can be used in subsequent filters.

### Configuration

| Name            | Type   | Description                                   | Required |
|-----------------|--------|-----------------------------------------------|----------|
| template        | string | template to create data, please refer the [template](https://github.com/megaease/easegress/blob/main/doc/reference/filters.md#template-of-builder-filters) for more information        | Yes       |
| dataKey         | string | key to store data        | Yes       |
| leftDelim       | string | left action delimiter of the template, default is `{{`  | No       |
| rightDelim      | string | right action delimiter of the template, default is `}}` | No       |

### Results

| Value                                                                       | Description                                        |
| --------------------------------------------------------------------------- | -------------------------------------------------- |
| buildErr                                                                    | Error happens when build the result.               |

### PR DESCRIPTION
**Pull request by caoshengdong**
resolves #803 
Add DataBuilder, tests for DataBuilder, and document for DataBuilder.

### CODE REVIEW

File: pkg/filters/builder/databuilder.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	"fmt"
+
+	"github.com/megaease/easegress/pkg/context"
+	"github.com/megaease/easegress/pkg/filters"
+	"github.com/megaease/easegress/pkg/logger"
+)
+
+const (
+	// DataBuilderKind is the kind of DataBuilder.
+	DataBuilderKind = "DataBuilderKind"
```
Comment (localvar) [inline pkg/filters/builder/databuilder.go:30]:
```suggestion
	DataBuilderKind = "DataBuilder"
```

File: doc/reference/filters.md [hunk 2]
```diff
@@ -947,6 +950,73 @@ filters:
...
 | result9                                                                     |
 
+## DataBuilder
+
+DataBuilder is used to manipulate and store data. The data from the previous
+filter can be transformed and stored in the context so that the data can be
+used in subsequent filters.
+
+For example, we can use DataBuilder to store token from login request, and
+use the token in subsequent requests.
```
Comment (localvar) [inline doc/reference/filters.md:960]:
```suggestion
use the token in subsequent filters.
```

File: doc/reference/filters.md [hunk 3]
```diff
@@ -947,6 +950,73 @@ filters:
...
+| dataKey         | string | key to store data        | Yes      |
+| leftDelim       | string | left action delimiter of the template, default is `{{`  | No       |
+| rightDelim      | string | right action delimiter of the template, default is `}}` | No       |
+
+
+### Results
+
+| Value                                                                       | Description                                        |
+| --------------------------------------------------------------------------- | -------------------------------------------------- |
+| buildErr                                                                    | Error happens when build the result.               |
```
Comment (localvar) [inline doc/reference/filters.md:1017]:
```suggestion
| buildErr                                                                    | Error happens when building the data.               |
```

File: pkg/filters/builder/databuilder.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	data, err := prepareBuilderData(ctx)
+
+	if err != nil {
+		logger.Warnf("prepareBuilderData failed: %v", err)
+		return resultBuildErr
+	}
+
+	var r interface{}
+	if err = rb.build(data, &r); err != nil {
+		msgFmt := "DataBuilder(%s): failed to build request info: %v"
```
Comment (localvar) [inline pkg/filters/builder/databuilder.go:117]:
```suggestion
		msgFmt := "DataBuilder(%s): failed to build data: %v"
```

File: doc/reference/filters.md [hunk 5]
```diff
@@ -947,6 +950,73 @@ filters:
 | ...                                                                         |
 | result9                                                                     |
 
+## DataBuilder
+
+DataBuilder is used to manipulate and store data. The data from the previous
+filter can be transformed and stored in the context so that the data can be
+used in subsequent filters.
+
+For example, we can use DataBuilder to store token from login request, and
```
Comment (localvar) [inline doc/reference/filters.md:959]: the example is correct, but the issue is that the request is never sent, please refer to other filters like `RequestBuilder` or `ResponseBuilder`.

btw, a completed example is not a must, it is fine to just show how to use it and how to use the data it generates.

File: pkg/filters/builder/databuilder.go (hunk 1 above)
Comment (caoshengdong) [reply pkg/filters/builder/databuilder.go]: Done

File: doc/reference/filters.md (hunk 2 above)
Comment (caoshengdong) [reply doc/reference/filters.md]: Done

File: doc/reference/filters.md (hunk 5 above)
Comment (caoshengdong) [reply doc/reference/filters.md]: Changed the sample to a simpler one.

**Review by localvar — APPROVED** [review]

**Review by xxx7xxxx — APPROVED** [review]

**Review by suchen-sci — APPROVED** [review]

**Pull request merged**